packages/key-utils/src/functions.ts

```diff
@@ -341,3 +351,82 @@ export const padLeft = (args: { data: string; size?: number; padString?: string
   const length = padString.length
   return padString.repeat((size - data.length) / length) + data
 }
+
+
+const uint8ArrayToHex = (uint8Array: Uint8Array): string =>
+  Array.from(uint8Array)
+    .map(byte => byte.toString(16).padStart(2, '0'))
+    .join('')
+
+const sanitizeHexString = (hexString: string): string => {
+  let cleanHex = hexString.replace(/[^0-9a-fA-F]/g, '')
+  if (cleanHex.length % 2 !== 0) {
+    cleanHex = '0' + cleanHex
+  }
+  return cleanHex
+}
+
+export const rawPublicKeyHexFromAsn1Der = (derKey: Uint8Array | string): string => {
+  let inputHex: string
+  if (typeof derKey === 'string') {
+    inputHex = sanitizeHexString(derKey)
+  } else {
+    inputHex = sanitizeHexString(uint8ArrayToHex(derKey))
+  }
+
+  logger.debug("Sanitized Input DER Key (Hex):", inputHex)
+
+  const derBytes = new Uint8Array(inputHex.match(/.{1,2}/g)!.map(byte => parseInt(byte, 16)))
+
+  if (derBytes[0] !== 0x30) {
+    throw new Error("Invalid DER encoding: Expected to start with sequence tag")
+  }
+
+  let index = 2
+  if (derBytes[1] & 0x80) {
+    const lengthBytesCount = derBytes[1] & 0x7F
+    index += lengthBytesCount
+  }
+
+  while (index < derBytes.length) {
+    if (derBytes[index] === 0x03) {
+      break
+    }
+    index++
+  }
+
+  if (index >= derBytes.length) {
+    throw new Error("Invalid DER encoding: Bit string not found")
+  }
+
+  // Skip over Bit String tag (0x03) and the length of the Bit String
+  index += 2
+
+  // Extract the number of unused bits in the Bit String
+  const unusedBits = derBytes[index]
+  logger.debug("Unused bits in Bit String:", unusedBits)
+  index++
+
+  // Extract the raw public key bytes
+  const rawPublicKeyBytes = derBytes.slice(index)
+  const rawPublicKeyHex = uint8ArrayToHex(rawPublicKeyBytes)
+
+  logger.debug("Raw public key starts at index:", index)
+  logger.debug("Output Raw Public Key (Hex):", rawPublicKeyHex)
+
+  // Determine the key type and validate length
+  if (rawPublicKeyHex.length === 130 || rawPublicKeyHex.length === 128) {
+    logger.debug("Detected uncompressed secp256k1 or secp256r1 key")
+    // If it starts with '04', it's the full uncompressed key. Otherwise, we assume it's missing the '04' prefix.
+    return rawPublicKeyHex.startsWith('04') ? rawPublicKeyHex : '04' + rawPublicKeyHex
+  } else if (rawPublicKeyHex.length === 66 || rawPublicKeyHex.length === 64) {
+    logger.debug("Detected compressed secp256k1 or secp256r1 key")
+    // If it starts with '02' or '03', it's the full compressed key. Otherwise, we assume it's missing the prefix.
+    return (rawPublicKeyHex.startsWith('02') || rawPublicKeyHex.startsWith('03')) ? rawPublicKeyHex : '02' + rawPublicKeyHex
```

Comment:
If it is missing the prefix then it would either be 04 for an uncompressed key, or 02 or 03 depending on whether the hex value is even or odd (or the other way around)

Reply:
rawPublicKeyHexFromAsn1Der was completely rewritten.